### PR TITLE
Move context creation and closing to tchannel server hooks

### PR DESCRIPTION
**Pull request by robskillington**
cc @xichen2020 @martin-mao @mmihic

### CODE REVIEW

File: network/server/httpjson/handlers.go [hunk 1]
```diff
@@ -108,23 +117,62 @@ func RegisterHandlers(mux *http.ServeMux, service interface{}, opts ServerOption
...
+				if len(values) > 0 {
+					headers[key] = values[0]
+				}
 			}
 
-			svc := reflect.ValueOf(service)
+			var in interface{}
+			if reqIn != nil {
+				in = reflect.New(reqIn.Elem()).Interface()
+				defer r.Body.Close()
```
Comment (xichen2020) [inline network/server/httpjson/handlers.go]: We should close the body regardless whether it's a POST or a GET

Comment (robskillington) [inline network/server/httpjson/handlers.go]: I'm pretty sure if you never open it you don't need to close it.

Comment (xichen2020) [inline network/server/httpjson/handlers.go]: hm..just realized this is a request body, in which case I don't think we need to close it in either case.

http://grokbase.com/t/gg/golang-nuts/138p15ng41/go-nuts-http-request-body-close

Comment (robskillington) [inline network/server/httpjson/handlers.go]: Good to know documentation still doesn't exist about this.  Cool I'll remove close.

**Comment by xichen2020** [comment]
LGTM

File: network/server/httpjson/handlers.go (hunk 1 above)
Comment (xichen2020) [inline network/server/httpjson/handlers.go]: Well yeah kind of..it's hidden in the comments about the `Body` field inside the `Request` definition:
```
        // Body is the request's body.
        //
        // For client requests a nil body means the request has no
        // body, such as a GET request. The HTTP Client's Transport
        // is responsible for calling the Close method.
        //
        // For server requests the Request Body is always non-nil
        // but will return EOF immediately when no body is present.
        // The Server will close the request body. The ServeHTTP
        // Handler does not need to.
        Body io.ReadCloser
```

**Pull request merged**